text/projects/Krzysztof_Michalik/Wee_LCP.tex

```diff
@@ -0,0 +1,179 @@
+
+Teraz możemy poczynić następujące obserwacje na temat $S$:
+\begin{enumerate}
+    \item Liczba bitów '1' w $S$ wynosi dokładnie $n$, ponieważ każdy element $DIFF_S[i]$ miał w zapisie dokładnie jedną jedynkę.
+    \item Liczba bitów '0' jest ograniczona od góry przez $n$ z faktu, ze liczba zer oznaczała różnicę pomiędzy kolejnymi elementami ciągu $a_i$, a suma tych różnic nie mogła przekraczać $n$, ponieważ ciąg $a_i$ był rosnący i przyjmował jedynie wartości z przedziału $[1,n]$.
+\end{enumerate}
+
+Oznacza to, że taki string $S$ zajmuje nie więcej niż $2n$ bitów swoim zapisem i jest to zapis tablicy $LCP$ o który nam chodziło.
+
+W tym momencie autor powołuje się na rezultat z innej pracy - strukturę, która dla danego stringa bitowego pozwala odpowiadać na zapytania:
+\begin{enumerate}
+    \item $rank_1(S, i) = a$, gdzie $a$ jest liczbą wystąpień bitu '1' w prefiksie $S[:i]$. Innymi słowy $rank_1$ zlicza wystąpienia '1' w $S[:i]$.
+    \item $select_1(S, i) = a$, gdzie $S[a] = $'1' oraz $rank_1(S, i-1) = i-1$. Innymi słowy $select_1$ zwraca indeks $i$-tego bitu '1' w $S$.
+    \item Analogicznie definiowane są $rank_0(S, i)$ oraz $select_0(S, i)$.
+\end{enumerate}
+
+Struktura z której korzystają wykorzystuje $O(\frac{n \log{\log{n}}}{\log{n}})$ bitów i pozwala odpowiadać na powyższe zapytania w $O(1)$. Oznacza to, że w całości potrzebujemy $2n + O(\frac{n \log{\log{n}}}{\log{n}})$ bitów, aby pomieścić string $S$ oraz tę strukturę. Jest to dokładnie pamięć o jaką nam chodziło.
+
+Pozostaje pokazać jak wyliczamy $LCP[i]$ korzystając z $S$, $S_A$ i $T$. Niech $SUF[j] = i$. Wzory na to są następujące:
+
+%Pokażemy, że $LCP[i] = rank_0(S, select_1(S, SUF[i])) - SUF[i] \leftrightarrow LCP[i] + SUF[i] = %rank_0(S, select_1(S, SUF[i])) \leftrightarrow LCP[i] + SUF^{-1}[j] = rank_0(S, select_1(S, SUF[i]))$
+%W tym celu zaobserwujemy następujące fakty:
+
+\begin{enumerate}
+    \item $LCP[i] = rank_0(S, select_1(S, A[i])) - A[i]$
+    \item $LCP[i] = select_1(S, A[i]) - 2A[i]$
+\end{enumerate}
+
+%This is because the select1-statement gives the position where the encoding for H[A[i]] ends in H,
+%and the rank0-statement counts the sum of the I[j]’s for 1 ≤ j ≤ A[i]. So subtracting the value
+%A[i], which has been “artificially” added to the LCP-array, yields the correct value.
+
+Pierwszy wzór jest poprawny, ponieważ $select_1(S, A[i])$ zwraca pozycję, gdzie kończy się kodowanie $LCP[S_A[i]]$ w $S$. Wtedy $rank_0()$ zlicza sumę $DIFF[j]$, dla $j \le A[i]$. Zatem odjęcie $A[i]$ (które było sztucznie dodane aby zagwarantować monotoniczność ciągu), zwróci rzeczywistą wartość $LCP[i]$.
```

Comment:
"$rank_0()$ zlicza sumę $DIFF[j]$" - for completeness please changeit to $rank_0(S, select_1(S, A[i]))$